### ISSUE TITLE
ZynAddSubFX can't create a new instrument bank

### ISSUE THREAD

**Issue by pikachuk** [LMMS/lmms#4642]
I'm posting this issue because this issue that was present since some times in LMMS and was thought fixed https://github.com/LMMS/lmms/issues/3886 still exists

the steps to reproduce this bug are pretty easy : 
 - Starting LMMS on Ubuntu (this bug also exists on the stable 1.1.3 build)
- Adding a ZynAddSubFX track to the project
- Creating a new bank

screenshot of the issue : 
![capture_decran_2018-10-02_10-44-28](https://user-images.githubusercontent.com/16335601/46379258-84cd9c80-c69e-11e8-94a7-2c9db580f80c.png)

**Comment by tresf** [comment]
So I'm not sure if this is just an old bug inside ZynAddSubFX, but when I add a debug line to the method which creates the bank folder, it's attempting to [write to a location called `~/banks`](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/zynaddsubfx/src/Misc/Config.cpp#L88) at [pricely this location](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/zynaddsubfx/src/Misc/Bank.cpp#L266) of code.

Although from a casual observer, the directory `~/banks` appears to be a user-writable location, I don't believe `~` is a valid directory as it is interpreted literally as a folder named with tilde character. 
 For home, I would (instead expect to see a [helper function](https://stackoverflow.com/a/26696759/3196753) for resolving to `$HOME`.

Attempts to fix this hard-coded value will initially fail unless `.zynaddsubfxXML.cfg` is purged because  Zyn always prefers the value from the config file [so as long as the first element isn't blank](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/zynaddsubfx/src/Misc/Config.cpp#L86).

A temporary workaround is the following:

```bash
sed -i -e 's/~\/banks/\/tmp/' ~/Documents/lmms/.zynaddsubfxXML.cfg
```

This will allow the banks to be written temporarily to `/tmp`, which is a writable directory.

**Comment by musikBear** [comment]
Actually zasfx 'banks' are nothing but a normal folder, with at least one zasfx preset inside. If you make *that* zasfx will understand it as a bank, and it will be accessible from both zasfx, and indeed lmms -at least in windows....

**Comment by tresf** [comment]
@musikBear thanks, but this is about the bug where the folder cannot be created on Linux.  When testing, Windows does not suffer this problem.

**Comment by tresf** [comment]
Apparently upstream [patched this `~`](https://github.com/zynaddsubfx/zynaddsubfx/commit/ca9ec78952fc51616a684203fc4cff2fe101b11f) behavior three years ago... :)

**Comment by fundamental** [comment]
User data should not be stored in /tmp

So, setting the default bank directory to be within /tmp would not be
advisable.

**Comment by tresf** [comment]
Pardon my lack of understanding, I didn't understand if this was something that must be kept persistent.  Assuming is is we'll have to find a portable home for it.  This means we'll have to write our own mkdir logic.  How is it done in modern zyn?  How does the plugin get past `$HOME/banks` not existing?

**Comment by tresf** [comment]
It appears `newbank()` does attempt to create the directory.  To be consistent with upstream, `$HOME/banks` is a quick fix with the disadvantage of not being portable (as well as being non-standard, user folders should always be identified by the application name).

```cpp
#ifdef WIN32
    if(mkdir(bankdir.c_str()) < 0)
#else
    if(mkdir(bankdir.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) < 0)
#endif
```

But it's still too late I feel.  LMMS [already attempts to setup](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/LocalZynAddSubFx.cpp#L183) a sane bank location (we set it to the presets location) as long as [one of them is empty](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/LocalZynAddSubFx.cpp#L181).  Are we too late?  We can attempt to overwrite the value of `~/banks` if it's found in the config but is this going to break the loading of banks that are built-in?  Is this a matter of needing TWO bank locations?  First being ones we ship, second being ones we create?

**Comment by tresf** [comment]
After re-reading the source, I think it's safe to simply add a call to void `LocalZynAddSubFx::setPresetDir(...)` from within `LocalZynAddSubFx::setLmmsWorkingDir(...)`. Why? Well:
* Zyn allows [100](https://github.com/LMMS/lmms/blob/28a522806bb1a670b460f0bfbf8d6147910dea72/plugins/zynaddsubfx/zynaddsubfx/src/Misc/Config.h#L29) bank directories.
* We already loop over the hard-coded ones before appending.
* We already have the `ConfigManager::inst()->userPresetsDir() + "/ZynAddSubFX"` value as part of the call to `LocalZynAddSubFx::setLmmsWorkingDir(...)`.

**Comment by tresf** [comment]
> It appears `newbank()` does attempt to create the directory.

Correction... no, it doesn't.  It DOES create the new bank directory, but not the parent directory. This means on systems without `$HOME/banks`, https://github.com/zynaddsubfx/zynaddsubfx/commit/ca9ec78952fc51616a684203fc4cff2fe101b11f still doesn't fix it.

So I'll ask one more time... @fundamental how does Zyn handle this? Is there a startup task in newer versions that we're missing that creates this directory?  How does `$HOME/banks` get created on a new system?  This portion of the code appears largely untouched, even in 3.0.

We can still handle this LMMS-side, but we'd be creating a user directory that's only ever used by Zyn.  I'd expect Zyn to have the code to create this somewhere since new banks aren't possible without it.

**Comment by tresf** [comment]
Ended up adding an additional `mkdir` which I still feels is an upstream bug.  @JohannesLorenz is investigating if 3.0 has this bug too, but in the meantime, we're patched.  This will be patched in 1.2.0-RC9 (or higher).  If it doesn't work, try deleting `~/lmms/.zynaddsubfxXML.cfg` and trying again.

Ideally, (long-term) we'd like to make these settings portable and store banks somewhere inside `<LMMS_WORKING_DIR>`, but for now -- as a stop-gap -- we'll at least allow the "New Bank" button to work as it would normally.
